### DISCUSSION TITLE
Maximum allowed value of ulimit -n

### DISCUSSION THREAD

**Discussion by cyrilarnould** [msys2/MINGW-packages#11674] [General]
I often perform grep -r searches, sometimes over very large directories. At some point, those searches fail with `too many open files`. I similarly run into this problem when I want to natively compile elisp packages for Emacs in parallel.

Searching online I've found that ulimit -n is responsible for this limitation. When I try to change the limit however, it fails saying `cannot modify limit: Too many open files`. I guess this probably also comes from cygwin. Curiously, I find very little discussion on this topic aside from people saying it's not possible, e.g. https://cygwin.com/pipermail/cygwin/2006-October/151485.html

It would be great to see this implemented, although there's probably a reason why this can't easily be changed. I'd be interested to find out why though. I'd also welcome any workarounds to my problems, it seems to me like I can't be the only one to regularly run into this.

**Comment by Compyx** [comment]
The issue with grep can be solved by using `find` with the -exec switch, for example:
`find src \( -name '*.c' -o -name '*.h' \) -exec grep '#include' {} \;`
will search for all .c/.h files under src/ and execute `grep` for each file found, one at a time.

As for the limit on open files, I wouldn't be surprised if that's hardcoded in Windows.

**Reply by cyrilarnould** [reply]
Thanks for the suggestion. I have two gripes with the solution:

- The output doesn't show which file the line was found in
- findutils only exists in the MSYS2 environment

Incidentally, I've been wondering why findutils hasn't been added to the MINGW-packages yet. Simply because no one's done it yet?

**Reply by Biswa96** [reply]
[findutils][1] is available as msys2 package because it requires posix functions which is provided my msys2 layers (forked from cygwin).

[1]: https://packages.msys2.org/base/findutils

**Reply by cyrilarnould** [reply]
> [findutils](https://packages.msys2.org/base/findutils) is available as msys2 package because it requires posix functions which is provided my msys2 layers (forked from cygwin).

Well yes, but native ports do exist, see GnuWin32 or ezwinports for example.

**Reply by 1480c1** [reply]
I'm not sure why, but I never thought to look into the fact that there is a `${MINGW_PACKAGE_PREFIX}-grep`.

Just to make sure, which `grep` are you using? This discussion was made in the MSYS2-packages repo instead of the MINGW-packages repo, so I might assume `msys/grep` rather than `${MINGW_PACKAGE_PREFIX}-grep`

**Reply by cyrilarnould** [reply]
Sorry for the confusion, I do use mingw-w64-x86_64-grep more. I thought the discussion more appropriate here though because I wanted to focus on the ulimit issue.

**Reply by 1480c1** [reply]
Are you able to reproduce the issue outside of bash while using the mingw-w64 based grep, say something like powershell? Can we try to isolate only grep outside the context of bash/cygwin to see if it's a Windows specific issue rather than a `ulimit` issue?

**Reply by cyrilarnould** [reply]
Huh. It never even occurred to me that this would be a Windows issue, but you're right. If I run /usr/bin/grep instead, I don't run into the problem.

I guess I'll create an issue in the MINGW-packages repository. Is it possible to delete this discussion? Doesn't seem necessary anymore.

**Reply by 1480c1** [reply]
I think @Biswa96 or @lazka would be able to migrate this to the other repo without needing to delete/recreate

**Reply by 1480c1** [reply]
But yeah, since you mentioned `${MINGW_PACKAGE_PREFIX}-grep` it would probably be unrelated to bash/cygwin since those binaries are supposed to be able to work even without using msys2, and it's provided terminals and shells

**Reply by 1480c1** [reply]
One thing you could try is see is how many files you are trying to grep? Maybe `find . -type f | wc -l` might be able to get the count.

https://superuser.com/questions/1356320/what-is-the-number-of-open-files-limits

That page might be related to this issue

**Reply by Biswa96** [reply]
Windows has also a limit of handles per process https://devblogs.microsoft.com/oldnewthing/20070718-00/?p=25963. Is there some sort of memory leaking?

**Reply by cyrilarnould** [reply]
I'm currently running some tests to see what the limitations are. The limit is definitely not 3200.

If I just grep -r in my Windows %APPDATA% folder the error occurs. It's 57'296 files and a total of 18'255 folders.

But if I create a test folder with 200'000 files, I don't get any issues.

**Reply by 1480c1** [reply]
Perhaps you need to simulate a ton of folders? I can only assume that a lot of the concurrently open files would be folders, but I haven't dug deep into grep's source code to see if it fails to close anything

**Reply by cyrilarnould** [reply]
Ok, I have a testcase. Should I describe it here or create a separate issue?

**Reply by 1480c1** [reply]
If you truly found an issue, an issue would probably be better

**Reply by cyrilarnould** [reply]
I actually only realized afterwards that the 200'000 files also were problematic in that I only get results for the first 100'000 files (but no `Too many open files` message). I've created two separate issues for the two problems: #11675 and  #11676.
